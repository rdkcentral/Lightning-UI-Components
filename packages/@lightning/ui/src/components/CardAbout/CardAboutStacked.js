import * as styles from './CardAboutStacked.styles';
import CardAbout from './CardAbout';
import { TextBox, withExtensions } from '@lightning/ui-core';

class CardAboutStacked extends CardAbout {
  static get __componentName() {
    return 'CardAboutStacked';
  }

  static get __themeStyle() {
    return styles;
  }

  static get properties() {
    const props = super.properties.filter(
      prop => !['iconRight', 'textRight', 'iconLeft', 'textLeft'].includes(prop)
    );
    return [...props, 'titleBottom', 'descriptionBottom'];
  }

  static get tags() {
    const tags = super.tags.filter(
      tag => !['LeftIconTextContainer', 'RightIconTextContainer'].includes(tag)
    );
    return [...tags, 'TitleBottom', 'DescriptionBottom'];
  }

  static _template() {
    return {
      ...super._template(),
      DescriptionBottom: { type: TextBox }
    };
  }

  _update() {
    super._updateTitleStyle();
    this._updateBottomTitle();
    super._update();
    this._updateBottomDescription();
  }

  _updateDescriptionPosition() {
    if (!this.description) {
      this._Title.alpha = 0;
      return;
    }
    this._Description.patch({
      x: this.style.paddingHorizontal,
      content: this.description,
      y:
        this._Title.y + this._Title.style.textStyle &&
        this._Title.style.textStyle.lineHeight +
          this.style.paddingVertical * 1.5,
      style: { textStyle: this.style.descriptionTextStyle },
      justify: 'flex-start'
    });
  }

  _updateBottomTitle() {
    let titleBottomObject = {};
    let y = this.style.paddingFirstLine;
    if (this.description) {
      const lineHeight =
        (this._Description.style.textStyle &&
          this._Description.style.textStyle.lineHeight) ||
        0;
      y = this._Description.y + lineHeight + this.style.paddingVertical * 2;
    }

    titleBottomObject = {
      content: this.titleBottom && this.titleBottom.toUpperCase(),
      x: this.style.paddingHorizontal,
      y,
      style: {
        textStyle: {
          ...this.style.titleTextStyle,
          wordWrapWidth: this._calculateTextWidth()
        }
      }
    };
    if (!this._TitleBottom) {
      titleBottomObject.type = TextBox;
    }
    this.patch({ TitleBottom: titleBottomObject });
  }

  _updateBottomDescription() {
    if (this.descriptionBottom) {
      this._DescriptionBottom.patch({
        content: this.descriptionBottom,
        alpha: this.titleBottom ? 1 : 0,
        x: this.style.paddingHorizontal,
        y:
          this._TitleBottom.style.textStyle.lineHeight +
          this._TitleBottom.y +
          this.style.paddingVertical,
        style: {
          textStyle: {
            ...this.style.descriptionTextStyle,
            wordWrapWidth: this._calculateTextWidth()
          }
        }
      });
    }
  }

  _updateContent() {
    return;
    // This method from CardAbout needs to be overridden
    // disabled because there is no RightIconTextContainer that is used in CardAbout
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return (
      this._announce || [
        this._Title && this._Title.announce,
        this._Description && this._Description.announce,
        this._TitleBottom && this._TitleBottom.announce,
        this._DescriptionBottom && this._DescriptionBottom.announce
      ]
    );
  }
}

export default withExtensions(CardAboutStacked);
