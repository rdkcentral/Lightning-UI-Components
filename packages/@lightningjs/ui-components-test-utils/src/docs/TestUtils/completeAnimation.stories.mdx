<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / completeAnimation" />

# completeAnimation

Returns a Promise that resolves once all animating properties have updated to their target value(s).

## Arguments

`completeAnimation(element, transitionProperties = [])`

| argument             | type              | default | description                                                 |
| -------------------- | ----------------- | ------- | ----------------------------------------------------------- |
| element              | Lightning.Element | -       | element with properties that transition value changes       |
| transitionProperties | array\|string     | []      | property name(s) that transition value changes on component |

<br />

##### Example: Testing element properties before and after their values have transitioned

```js
import lng from '@lightningjs/core';
import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';

class Example extends lng.Component {
  static _template() {
    return {
      TransitioningRect: {
        rect: true,
        color: 0xffffffff,
        w: 100,
        h: 100,
        x: 0,
        y: 0
      }
    };
  }

  _updatePosition() {
    this.tag('TransitioningRect').setSmooth('x', this.axisPosition);
    this.tag('TransitioningRect').setSmooth('y', this.axisPosition);
  }

  set axisPosition(value) {
    this._axisPosition = value;
    this._updatePosition();
  }

  get axisPosition() {
    return this._axisPosition;
  }
}

// generate a createComponent function to use in tests
const createExampleComponent = makeCreateComponent(Example);

// Ex. await the transitions of x and y values to finish before testing the result
it('should transition the x value of the component', async () => {
  const [example] = createExampleComponent();
  expect(example.tag('TransitioningRect').x).toBe(0);
  expect(example.tag('TransitioningRect').y).toBe(0);

  example.axisPosition = 50;
  await completeAnimation(example.tag('TransitioningRect'), ['x', 'y']);

  expect(example.tag('TransitioningRect').x).toBe(50);
  expect(example.tag('TransitioningRect').y).toBe(50);
});
```
