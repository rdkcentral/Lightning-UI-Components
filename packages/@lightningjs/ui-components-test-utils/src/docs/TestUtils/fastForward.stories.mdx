<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / fastForward" />

# fastForward

Force all running Lightning Transitions on one or more Lightning Elements to finish and update the transitioning property to its target value immediately.

This function is similar to [completeAnimation](/story/unit-testing-test-utils-completeanimation--page) in that it allows testing transitioning properties. Unlike `completeAnimation`, `fastForward` is not asynchronous, and does result in all running transitions to complete. If a test requires awaiting specific properties to finish transitioning, `completeAnimation` is the recommended utility function. The `fastForward` function forces transitions to finish, but does not wait for the template to update before proceeding through the test function. To ensure the template has updated following a call to `fastForward`, invoke [`testRenderer.update()`](/story/unit-testing-test-renderer-overview--page).
// TODO: add link directly to testRenderer.update() docs once those are written

## Arguments

`fastForward(elements)`

| argument | type                                           | default | description                                   |
| -------- | ---------------------------------------------- | ------- | --------------------------------------------- |
| elements | Array<Lightning.Element\> \| Lightning.Element | -       | element(s) with transitioning property values |

<br />

##### Example: Forcing running Transitions to finish before testing the result

```js
import lng from '@lightningjs/core';
import {
  makeCreateComponent,
  fastForward
} from '@lightningjs/ui-components-test-utils';

class Example extends lng.Component {
  static _template() {
    const square = {
      rect: true,
      w: 100,
      h: 100,
      x: 0,
      y: 0,
      color: 0xeeeeeeee
    };
    return {
      TransitioningRectA: square,
      TransitioningRectB: square
    };
  }

  set axisPosition(value) {
    this._axisPosition = value;
    this.tag('TransitioningRectA').setSmooth('x', value);
    this.tag('TransitioningRectB').setSmooth('y', value);
  }

  get axisPosition() {
    return this._axisPosition;
  }
}

// generate a createComponent function to use in tests
const createExampleComponent = makeCreateComponent(Example);

// Ex. force all running transitions to complete, then test the result
it('should transition the x value of the component', () => {
  const [example, testRenderer] = createExampleComponent();
  expect(example.tag('TransitioningRectA').x).toBe(0);
  expect(example.tag('TransitioningRectB').y).toBe(0);

  //
  example.axisPosition = 50;
  const rectA = example.tag('TransitioningRectA');
  const rectB = example.tag('TransitioningRectB');
  fastForward([rectA, rectB]);
  testRenderer.update();

  expect(example.tag('TransitioningRectA').x).toBe(50);
  expect(example.tag('TransitioningRectB').y).toBe(50);
});
```
