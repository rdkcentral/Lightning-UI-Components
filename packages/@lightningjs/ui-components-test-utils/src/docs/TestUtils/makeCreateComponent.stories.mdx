<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Docs / Unit Testing / Test Utils / makeCreateComponent " />

# makeCreateComponent

Returns a function, referred to as a `createComponent` function, which is used to generate an instance of a component to run unit test cases against.

## Arguments

`makeCreateComponent(component, defaultConfig, defaultOptions)`

| argument       | type          | default                                  | description                                                                                                                |
| -------------- | ------------- | ---------------------------------------- | -------------------------------------------------------------------------------------------------------------------------- |
| component      | lng.Component | the component to generate an instance of | component to render an instance of                                                                                         |
| defaultConfig  | object        | {}                                       | properties that should be applied to the component by default                                                              |
| defaultOptions | object        | {}                                       | default options used when creating the component instance (these options are passed to the `TestRenderer.create` function) |

### `createComponent`

Returns an array of 2 elements:

1. an instance of the component to be tested against (generated by `testRenderer.getInstance()`)
2. a [testRenderer](/story/unit-testing-test-renderer-create--page#testrenderer-functions) object generated using the `component` argument and a merged object created from `defaultOptions` and `options`

#### Arguments

| argument | type   | default | description                                                                                                                                                                |
| -------- | ------ | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| config   | object | {}      | properties that should be applied to the component. This object is spread over the `defaultConfig` object.                                                                 |
| options  | object | {}      | options used when creating the component instance (these options are passed to the `TestRenderer.create` function). This object is spread over the `defaultOptions` object |

<br />

##### Example: Generating a `makeComponent` function and test component

```js
import lng from '@lightningjs/core';
import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';

// Lightning component to write unit tests for
class Example extends lng.Component {
  static _template() {
    return {
      Example: {
        text: { text: '' }
      }
    };
  }

  set exampleProp(value) {
    this._exampleProp = value;
    this.tag('Example').text.text = value;
  }

  get exampleProp() {
    return this._exampleProp;
  }
}

// unit tests for Example component

// generate a createComponent function to use in tests
const defaultConfig = { exampleProp: 'foo' };
const defaultOptions = { focused: false };
const createExampleComponent = makeCreateComponent(
  Example,
  defaultConfig,
  defaultOptions
);

// Ex. An instance of the component rendered with the properties defined in the defaultConfig and defaultOptions object
it('should render the exampleProp value as text', () => {
  const [example] = createExampleComponent();

  expect(example.exampleProp).toBe('foo');
  expect(example.tag('Example').text.text).toBe('foo');
  expect(example.hasFocus()).toBe(false);
});

// Ex. An instance of the component rendered with the properties defined in the config object
it('should support updating the text', () => {
  // overwrite the value of exampleProp defined in defaultConfig
  const [example] = createExampleComponent({
    exampleProp: 'bar'
  });

  expect(example.exampleProp).toBe('bar');
  expect(example.tag('Example').text.text).toBe('bar');
});

// Ex. An instance of the component rendered with options defined in the options object
it('should support rendering the component in a focused state', () => {
  // overwrite the value of focused defined in defaultOptions
  const [example] = createExampleComponent({}, { focused: true });

  expect(example.hasFocus()).toBe(true);
});

// Ex. Using a function provided by the testRenderer object (focus)
it('should allow focusing on the component', () => {
  const [example, testRenderer] = createExampleComponent();

  // focused was set to false in defaultOptions, so the componet will initially render without focus
  expect(example.hasFocus()).toBe(false);

  testRenderer.focus();

  expect(example.hasFocus()).toBe(true);
});
```

### spyOnMethods

A utility which enables running component methods asynchronously.

The `options` argument passed to the `createComponent` function accepts a field called `spyOnMethods`.
The value for `spyOnMethods` can be an array of method names to generate promises from. When one of those methods is invoked in the component it's associated promise will resolve, then the actual method will be invoked. The promises generated from the passed in methods are given the name `_${methodName}SpyPromise`.

<br />

##### Example: Spying on component methods and awaiting their execution before making assertions on their side effects

```js
class MethodSpiesExample extends Base {
  static get __componentName() {
    return 'MethodSpiesExample';
  }

  static _template() {
    return {
      Title: {
        type: TextBox
      },
      Subtitle: {
        type: TextBox,
        y: 100
      }
    };
  }

  static get tags() {
    return ['Title', 'Subtitle'];
  }

  static get properties() {
    return ['title', 'subtitle'];
  }

  _construct() {
    super._construct();
    this._title = 'title';
    this._subtitle = 'subtitle';
  }

  _update() {
    this._updateTitle();
    this._updateSubtitle();
  }

  _updateTitle() {
    this._Title.content = this.title;
  }
  _updateSubtitle() {
    this._Subtitle.content = this.subtitle;
  }
}

const createComponent = makeCreateComponent(MethodSpies);

it('should update the title and subtitle', async () => {
  const [methodSpiesComponent, testRenderer] = createComponent(
    {},
    { spyOnMethods: ['_update'] }
  );

  // wait for _update to be called on the initial render before proceeding
  await methodSpiesComponent.__updateSpyPromise;

  expect(methodSpiesComponent._Title.content).toBe('title');
  expect(methodSpiesComponent._Subtitle.content).toBe('subtitle');

  const newTitle = 'new title';
  const newSubtitle = 'new subtitle';
  methodSpiesComponent.patch({
    title: newTitle,
    subtitle: newSubtitle
  });

  // wait for _update to be called in response to values in the properties array changing
  await methodSpiesComponent.__updateSpyPromise;

  expect(methodSpiesComponent._Title.content).toBe(newTitle);
  expect(methodSpiesComponent._Subtitle.content).toBe(newSubtitle);
});
```

**Note: `spyOnMethods` can only be set in in the `options` object of `createComponent`. It can not be assigned to the `defaultOptions` object in `makeCreateComponent`**
