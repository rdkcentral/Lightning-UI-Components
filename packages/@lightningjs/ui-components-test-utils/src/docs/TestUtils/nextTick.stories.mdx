<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / nextTick" />

# nextTick

Creates a Promise that resolves after a defined amount of time. If no amount of time is specified, the Promise will resolve immediately.

## Arguments

`nextTick(wait = 0)`

| argument | type   | default | description                                                           |
| -------- | ------ | ------- | --------------------------------------------------------------------- |
| wait     | number | 0       | amount of time (in milliseconds) before the returned Promise resolves |
