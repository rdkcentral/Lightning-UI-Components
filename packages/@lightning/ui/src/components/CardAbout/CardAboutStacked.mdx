import { Canvas, Story } from '@storybook/addon-docs';
import CardAboutStacked from './CardAboutStacked.js';

# CardAboutStacked

This component is used to display two titles with content below to create a stacked effect.

## Source

https://github.comcast.com/Lightning/lightning-ui/blob/develop/packages/@lightning/ui/src/components/CardAbout/CardAboutStacked.js

## Usage

`CardAboutStacked` takes in a title and a titleBottom to create a stacked effect. contentBottom is the inline content that can be passed in to go below the title, and descriptionBottom can be passed in to go below titleBottom.
`CardAboutStacked` extends [Card](?path=/docs/patterns-card--base)

```js
import { CardAboutStacked } from '@lightning/ui';

class Basic extends lng.Component {
  static _template() {
    return {
      CardAboutStacked: {
        type: CardAboutStacked,
        title: 'Audio',
        titleBottom: 'Title for the bottom part of the card'
      }
    };
  }
}
```

<Canvas>
  <Story id="tiles-cards-cardaboutstacked--stacked" />
</Canvas>

## API

### Properties

| name              | type   | required | default   | description                              |
| ----------------- | ------ | -------- | --------- | ---------------------------------------- |
| description       | object | false    | undefined | inline content in the middle of the card |
| titleBottom       | string | false    | undefined | title on the bottom half of the card     |
| descriptionBottom | string | false    | undefined | bottom description                       |

### Style Properties

| name                 | type             | description                      |
| -------------------- | ---------------- | -------------------------------- |
| descriptionTextStyle | string \| object | text style for the description   |
| paddingVertical      | string           | padding vertically between lines |
