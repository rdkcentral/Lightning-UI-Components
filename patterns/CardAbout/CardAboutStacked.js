import * as styles from './CardAboutStacked.styles';
import { withExtensions } from '../../mixins';
import CardAbout from './CardAbout';
import TextBox from '../../elements/TextBox';

class CardAboutStacked extends CardAbout {
  static get __componentName() {
    return 'CardAboutStacked';
  }

  static get __themeStyle() {
    return styles;
  }

  static get properties() {
    const props = super.properties.filter(
      prop => !['iconRight', 'textRight', 'iconLeft', 'textLeft'].includes(prop)
    );
    return [...props, 'titleBottom', 'descriptionBottom'];
  }

  static get tags() {
    const tags = super.tags.filter(
      tag => !['LeftIconTextContainer', 'RightIconTextContainer'].includes(tag)
    );
    return [...tags, 'TitleBottom', 'DescriptionBottom'];
  }

  static _template() {
    return {
      ...super._template(),
      DescriptionBottom: { type: TextBox }
    };
  }

  _update() {
    super._updateTitleStyle();
    this._updateBottomTitle();
    super._update();
    this._updateBottomDescription();
  }

  _updateDescriptionPosition() {
    if (!this.description) {
      this._Title.alpha = 0;
      return;
    }
    this._Description.patch({
      x: this.style.paddingHorizontal,
      content: this.description,
      y:
        this._Title.y + this._Title.textStyle &&
        this._Title.textStyle.lineHeight + this.style.paddingVertical * 1.5,
      textStyle: this.style.descriptionTextStyle
    });
  }

  _updateBottomTitle() {
    let titleBottomObject = {};
    let y = this.style.paddingFirstLine;
    if (this.description) {
      const lineHeight =
        (this._Description.style.textStyle &&
          this._Description.style.textStyle.lineHeight) ||
        0;
      y = this._Description.y + lineHeight + this.style.paddingVertical * 2;
    }

    titleBottomObject = {
      content: this.titleBottom && this.titleBottom.toUpperCase(),
      x: this.style.paddingHorizontal,
      y,
      wordWrapWidth: this._calculateTextWidth(),
      textStyle: this.style.titleTextStyle
    };
    if (!this._TitleBottom) {
      titleBottomObject.type = TextBox;
    }
    this.patch({ TitleBottom: titleBottomObject });
  }

  _updateBottomDescription() {
    if (this.descriptionBottom) {
      this._DescriptionBottom.patch({
        content: this.descriptionBottom,
        alpha: this.titleBottom ? 1 : 0,
        x: this.style.paddingHorizontal,
        y:
          this._TitleBottom.textStyle.lineHeight +
          this._TitleBottom.y +
          this.style.paddingVertical,
        wordWrapWidth: this._calculateTextWidth(),
        textStyle: this.style.descriptionTextStyle
      });
    }
  }

  _updateContent() {
    return;
    // This method from CardAbout needs to be overridden
    // disabled because there is no RightIconTextContainer that is used in CardAbout
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return (
      this._announce || [
        this._Title && this._Title.announce,
        this._Description && this._Description.announce,
        this._TitleBottom && this._TitleBottom.announce,
        this._DescriptionBottom && this._DescriptionBottom.announce
      ]
    );
  }
}

export default withExtensions(CardAboutStacked);
