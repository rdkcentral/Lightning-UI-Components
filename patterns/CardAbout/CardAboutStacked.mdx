import { Canvas, Story } from '@storybook/addon-docs';
import CardAboutStacked from './CardAboutStacked.js';

# CardAboutStacked

This component is used to display two titles with content below to create a stacked effect.

## Usage

`CardAboutStacked` takes in a title and a titleBottom to create a stacked effect. contentBottom is the inline content that can be passed in to go below the title, and descriptionBottom can be passed in to go below titleBottom.
`CardAboutStacked` extends [Card](?path=/docs/patterns-card--base)

```js
import { CardAboutStacked } from '@lightning/ui';

class Basic extends lng.Component {
  static _template() {
    return {
      CardAboutStacked: {
        type: CardAboutStacked,
        title: 'Audio',
        titleBottom: 'Title for the bottom part of the card'
      }
    };
  }
}
```

<Canvas>
  <Story id="patterns-CardAboutStacked--about-stacked-cards" />
</Canvas>

## API

### Properties

| name              | type   | required | default | description                              |
| ----------------- | ------ | -------- | ------- | ---------------------------------------- |
| description       | object | false    | -       | inline content in the middle of the card |
| titleBottom       | string | false    | -       | title on the bottom half of the card     |
| descriptionBottom | string | false    | -       | bottom description                       |

### Style Properties

| name                      | type   | description                      |
| ------------------------- | ------ | -------------------------------- |
| descriptionTextProperties | object | text style of the description    |
| paddingVertical           | string | padding vertically between lines |
