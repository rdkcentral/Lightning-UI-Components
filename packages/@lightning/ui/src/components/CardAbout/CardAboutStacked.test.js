import TestUtils from '@lightning/ui-core/test';
import CardAboutStacked from './CardAboutStacked';

const createCardAboutStackedComponent =
  TestUtils.makeCreateComponent(CardAboutStacked);

describe('cardAboutStacked', () => {
  let cardAboutStacked, testRenderer;

  beforeEach(() => {
    [cardAboutStacked, testRenderer] = createCardAboutStackedComponent({
      title: 'someString',
      titleBottom: 'some string'
    });
  });

  afterEach(() => {
    cardAboutStacked = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the title, description, and bottom title and bottle description text', () => {
    const title = 'Title';
    const description = 'Description';
    const titleBottom = 'Title Bottom';
    const descriptionBottom = 'Description Bottom';
    cardAboutStacked.patch({
      title,
      description,
      titleBottom,
      descriptionBottom
    });
    testRenderer.forceAllUpdates();
    expect(cardAboutStacked.announce).toEqual([
      title.toUpperCase(),
      description,
      titleBottom.toUpperCase(),
      descriptionBottom
    ]);
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    cardAboutStacked.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(cardAboutStacked.announce).toBe(overrideString);
  });

  it('has the correct static props getter', () => {
    expect(cardAboutStacked.constructor.properties).toMatchObject([
      'title',
      'description',
      'titleBottom',
      'descriptionBottom'
    ]);
  });

  it('has the correct static tags getter', () => {
    expect(cardAboutStacked.constructor.tags).toMatchObject([
      'Background',
      'Title',
      'Description',
      'TitleBottom',
      'DescriptionBottom'
    ]);
  });

  // TODO: Should Stacked Be still using LeftIconTextContainer if there is not an icon?
  it.skip('should update Container', () => {
    cardAboutStacked.description = [
      { badge: 'badge' },
      { badge: 'badge' },
      { badge: 'badge' },
      { badge: 'badge' },
      { badge: 'badge' }
    ];
    cardAboutStacked.descriptionBottom = 'testing';
    cardAboutStacked.titleBottom = 'testing card';
    testRenderer.forceAllUpdates();
    expect(cardAboutStacked._TitleBottom.y).toEqual(
      cardAboutStacked._Description.y +
        cardAboutStacked._Description.style.textStyle.lineHeight +
        cardAboutStacked.style.paddingVertical * 2
    );
    expect(cardAboutStacked._Description.x).toEqual(
      cardAboutStacked.style.paddingHorizontal
    );
    expect(cardAboutStacked._Description.y).toEqual(
      cardAboutStacked._Title.y +
        cardAboutStacked._Title.textStyle.lineHeight +
        cardAboutStacked.style.paddingVertical * 1.5
    );
    expect(cardAboutStacked._DescriptionBottom.y).toEqual(
      cardAboutStacked._TitleBottom.textStyle.lineHeight +
        cardAboutStacked._TitleBottom.y +
        cardAboutStacked.style.paddingVertical
    );
  });

  it('should update Bottom text', () => {
    cardAboutStacked.titleBottom = 'Languages';

    testRenderer.forceAllUpdates();
    expect(cardAboutStacked._TitleBottom.content).toEqual('LANGUAGES');
    expect(cardAboutStacked._TitleBottom.x).toEqual(
      cardAboutStacked.style.paddingHorizontal
    );
    expect(cardAboutStacked._TitleBottom.y).toEqual(
      cardAboutStacked.description
        ? cardAboutStacked._Description.y +
            cardAboutStacked._Description.style.textStyle.lineHeight +
            cardAboutStacked.style.paddingVertical * 2
        : cardAboutStacked.style.paddingFirstLine
    );
    expect(cardAboutStacked._TitleBottom.textStyle).toEqual(
      cardAboutStacked.style.titleTextStyle
    );
  });

  it('should update Bottom text', () => {
    cardAboutStacked.titleBottom = 'Languages';

    testRenderer.forceAllUpdates();
    expect(cardAboutStacked._TitleBottom.content).toEqual('LANGUAGES');
    expect(cardAboutStacked._TitleBottom.x).toEqual(
      cardAboutStacked.style.paddingHorizontal
    );
    expect(cardAboutStacked._TitleBottom.y).toEqual(
      cardAboutStacked.description
        ? cardAboutStacked._Description.y +
            cardAboutStacked._Description.style.textStyle.lineHeight +
            cardAboutStacked.style.paddingVertical * 2
        : cardAboutStacked.style.paddingFirstLine
    );
    expect(cardAboutStacked._TitleBottom.textStyle).toEqual(
      cardAboutStacked.style.titleTextStyle
    );
  });
  it('should update position of Description', () => {
    cardAboutStacked.descriptionBottom = 'Description';
    cardAboutStacked.titleBottom = 'Languages';
    testRenderer.forceAllUpdates();
    expect(cardAboutStacked._DescriptionBottom.x).toEqual(
      cardAboutStacked.style.paddingHorizontal
    );
    expect(cardAboutStacked._DescriptionBottom.y).toEqual(
      cardAboutStacked.description
        ? cardAboutStacked._TitleBottom.y +
            cardAboutStacked.style.paddingVertical * 2
        : cardAboutStacked._TitleBottom.y +
            cardAboutStacked.style.paddingVertical * 3
    );
    expect(cardAboutStacked._DescriptionBottom.textStyle).toEqual(
      cardAboutStacked.style.descriptionTextStyle
    );
  });
});
