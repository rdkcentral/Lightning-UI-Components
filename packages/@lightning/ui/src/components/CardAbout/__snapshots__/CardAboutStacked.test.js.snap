// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cardAboutStacked renders 1`] = `
{
  "Component": {
    "Background": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Background",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Background",
      ],
      "texture": {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "Title": {
      "active": false,
      "alpha": 0,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": false,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Title",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": [
        "Title",
      ],
      "type": "TextBox",
      "visible": true,
      "w": 0,
      "x": 24,
      "y": 20,
      "zIndex": 0,
    },
  },
}
`;
